modules/agent_engine/main.tf

```
@@ -0,0 +1,212 @@
+        ? { 1 = 1 }
+        : {}
+      )
+
+      content {
+        dynamic "env" {
+          for_each = var.environment_variables
+
+          content {
+            name  = env.key
+            value = env.value
+          }
+        }
+
+        dynamic "secret_env" {
+          for_each = var.secret_environment_variables
+
+          content {
+            name = secret_env.key
+
+            secret_ref {
+              secret  = secret_env.value.secret_id
+              version = secret_env.value.version
+            }
+          }
+        }
+      }
+    }
+
+    package_spec {
+      python_version           = var.python_version
+      dependency_files_gcs_uri = "gs://${local.bucket_name}/${google_storage_bucket_object.dependencies.name}"
+      requirements_gcs_uri     = "gs://${local.bucket_name}/${google_storage_bucket_object.requirements.name}"
+      pickle_object_gcs_uri = (
+        var.generate_pickle
+        ? "gs://${local.bucket_name}/${google_storage_bucket_object.pickle_from_src[0].name}"
+        : "gs://${local.bucket_name}/${google_storage_bucket_object.pickle[0].name}"
+      )
+    }
+  }
+}
+
+# Eventual consistency issue.
+# AE doesn't retry (yet) the deployment if bindings are still not active.
+resource "time_sleep" "wait_5_minutes" {
+  create_duration = "5m"
+
+  depends_on = [
+    google_project_iam_member.default
+  ]
+}
+
+resource "google_storage_bucket" "default" {
+  count                       = var.bucket_config.create ? 1 : 0
+  name                        = coalesce(var.bucket_config.name, var.name)
+  project                     = local.project_id
+  location                    = local.location
+  uniform_bucket_level_access = var.bucket_config.uniform_bucket_level_access
+  force_destroy               = !var.bucket_config.deletion_protection
+}
+
+resource "null_resource" "default" {
+  count = var.generate_pickle ? 1 : 0
+
+  provisioner "local-exec" {
+    command = join(" ", [
+      "python",
+      "./tools/serialize_agent.py",
+      "${var.source_files.path}/${var.source_files.pickle_src}",
+      "--output-file ${var.source_files.path}/${var.source_files.pickle_out}",
+      "--variable-name ${var.source_files.pickle_src_var_name}"
+    ])
+  }
+}
+
+resource "google_storage_bucket_object" "dependencies" {
+  name           = "dependencies.tar.gz"
+  bucket         = local.bucket_name
+  source         = "${var.source_files.path}/${var.source_files.dependencies}"
+  source_md5hash = filemd5("${var.source_files.path}/${var.source_files.dependencies}")
+}
+
+resource "google_storage_bucket_object" "pickle_from_src" {
+  count          = var.generate_pickle ? 1 : 0
+  name           = "pickle.pkl"
+  bucket         = local.bucket_name
+  source         = "${var.source_files.path}/${var.source_files.pickle_out}"
+  source_md5hash = filemd5("${var.source_files.path}/${var.source_files.pickle_out}")
+
+  depends_on = [
+    null_resource.default
+  ]
+}
```

```suggestion
resource "local_file" "pickle_out" {
  filename = "${var.source_files.path}/${var.source_files.pickle_out}"
  depends_on = [ null_resource.default ]
}

resource "google_storage_bucket_object" "pickle_from_src" {
  count          = var.generate_pickle ? 1 : 0
  name           = "pickle.pkl"
  bucket         = local.bucket_name
  source         = "${var.source_files.path}/${var.source_files.pickle_out}"
  source_md5hash = local_file.pickle_out.content_md5
}
```

You need to create a separate `local_file`, otherwise terraform will call `filemd5()` during plan, and during apply this value will change, as you will regenerate the file. This is how you get inconsistent plan provider error (which in this case, is actually user error).